src/openrct2/ride/Ride.cpp

```diff
@@ -3134,136 +3100,106 @@ void ride_measurements_update()
+                    }
                 }
             }
         }
     }
 }
 
-static rct_ride_measurement* ride_get_existing_measurement(ride_id_t rideIndex)
-{
-    for (int32_t i = 0; i < MAX_RIDE_MEASUREMENTS; i++)
-    {
-        rct_ride_measurement* measurement = get_ride_measurement(i);
-        if (measurement->ride_index == rideIndex)
-            return measurement;
-    }
-
-    return nullptr;
-}
-
-static int32_t ride_get_free_measurement()
-{
-    for (int32_t i = 0; i < MAX_RIDE_MEASUREMENTS; i++)
-    {
-        rct_ride_measurement* measurement = get_ride_measurement(i);
-        if (measurement->ride_index == RIDE_ID_NULL)
-            return i;
-    }
-
-    return -1;
-}
-
 /**
- *
- *  rct2: 0x006B66D9
+ * If there are more than the threshold of allowed ride measurements, free the non-LRU one.
  */
-rct_ride_measurement* ride_get_measurement(Ride* ride, rct_string_id* message)
+static void ride_free_old_measurements()
 {
-    // Check if ride type supports data logging
-    if (!ride_type_has_flag(ride->type, RIDE_TYPE_FLAG_HAS_DATA_LOGGING))
-    {
-        if (message != nullptr)
-            *message = STR_DATA_LOGGING_NOT_AVAILABLE_FOR_THIS_TYPE_OF_RIDE;
-        return nullptr;
-    }
-
-    // Check if a measurement already exists for this ride
-    rct_ride_measurement* measurement = ride_get_existing_measurement(ride->id);
-    if (measurement == nullptr)
+    size_t numRideMeasurements;
+    do
     {
-        // Find a free measurement
-        int32_t i = ride_get_free_measurement();
-        if (i == -1)
+        ride_id_t i{};
+        Ride* ride{};
+        Ride* lruRide{};
+        numRideMeasurements = 0;
+        FOR_ALL_RIDES (i, ride)
         {
-            // Use last recently used measurement for some other ride
-            int32_t lruIndex = 0;
-            uint32_t lruTicks = 0xFFFFFFFF;
-            for (i = 0; i < MAX_RIDE_MEASUREMENTS; i++)
+            if (ride->measurement != nullptr)
             {
-                measurement = get_ride_measurement(i);
-
-                if (measurement->last_use_tick <= lruTicks)
+                if (lruRide == nullptr || ride->measurement->last_use_tick > lruRide->measurement->last_use_tick)
                 {
-                    lruTicks = measurement->last_use_tick;
-                    lruIndex = i;
+                    lruRide = ride;
                 }
+                numRideMeasurements++;
             }
-
-            i = lruIndex;
-            measurement = get_ride_measurement(i);
-            get_ride(measurement->ride_index)->measurement_index = 255;
         }
-        else
+        if (numRideMeasurements > MAX_RIDE_MEASUREMENTS && lruRide != nullptr)
         {
-            measurement = get_ride_measurement(i);
+            lruRide->measurement = {};
+            numRideMeasurements--;
         }
+    } while (numRideMeasurements > MAX_RIDE_MEASUREMENTS);
+}
+
+std::pair<RideMeasurement*, rct_string_id> ride_get_measurement(Ride* ride)
```

Comment:
Shouldn't this become a ride member function.

Reply:
I couldn't decide at the time whether it should be or not.